### DISCUSSION TITLE
pxrUsdReferenceAssembly Equivalent for Maya USD

### DISCUSSION THREAD

**Discussion by jonahjnewton** [Autodesk/maya-usd#2394] [General]
Hey,

I'm trying to create USD assembly reference nodes in Maya 2022 using cmds.assembly, however as pxrUsd is no longer used in Maya 2022, I am not sure what node type to use instead of pxrUsdReferenceAssembly. Maya's assemblyReference only seems to reference Maya scenes with assembly definitions, not USDs.

I managed to make a workaround by using proxyShapes, however my company's publish workflow requires cmds.assembly nodes, and proxyShapes do not have the assembly type.

Any help on this would be very appreciated.

Kind regards,
Jonah

**Comment by santosg87** [comment]
Hi @jjonahlowercase our framework on Maya 2023 _should_ have the needed nodes, as some of the comds you're looking for are deprecated, if I recall correctly. 

I will go ahead and move this thread to a discussion, and maybe we can get some answers for your case :)

**Reply by ppt-adsk** [reply]
Hi @jjonahlowercase, thanks for writing.  The Pixar plugin and its scene assembly nodes will not be further developed by Autodesk.  We are directing all our energy on the Autodesk plugin in the same open-source, Github repository.  Rather than focussing on the solution (which currently for you is to use the Pixar plugin and scene assembly nodes), we would rather like to know what workflows you are using in the Pixar plugin.  In Maya 2022, a lot of set dressing workflows can be addressed by using the Autodesk plugin proxy shape and native editing of USD data.  In Maya 2023, the Autodesk plugin will soon have functionality to allow you to edit Maya rigs and cache them to USD, all as part of the USD scene hierarchy, to address shot-based animation workflows.  Are these of interest to you?  Are you required to stay on Maya 2022, or is Maya 2023 an option?  Requiring the solution to working with USD in Maya to use scene assembly nodes and the Pixar plugin will prevent you from using these new features in the Autodesk plugin.  Let us know how we can best help you in your technology choices.  Cheers.

**Reply by jonahjnewton** [reply]
Thanks for replying! We currently do work in a shot-based workflow, we've been working in Maya 2019 up until now utilising pxrUsdReferenceAssembly nodes to reference our sets into shots. I totally understand what you're saying, and it sounds like we will need to rewrite our publishing pipeline to export with more native Maya USD functions, Maya 2022 is just what was out at the time when we were looking to move on from Maya 2019 but I don't see why we wouldn't be able to move up to Maya 2023 instead. We definitely don't want to miss out on new features but making sure it works with our pipeline is super important as well.

**Reply by jonahjnewton** [reply]
Hey, back again to ask further on this. 

It seems the main issue with our workflow right now is that to query edits to USD data when publishing a shot, we use the MayaAPI EditUtils script (the getEditsThatAffect method specifically) to query edits made to set assembly nodes (however this MayaAPI tool **requires** assembly nodes to be used). 

If assembly nodes are no longer supported, is there currently/are there any plans to have a MayaUSD alternative to query edits on a proxy shape? If this already exists, I'd love to learn more.

Thanks!

**Reply by mdubrois** [reply]
You can use the sessionLayer linked to the ProxyShape.

**Reply by jonahjnewton** [reply]
Hi, getting back to this.

I understand that I can use the sessionLayer as the edit target for edits to a USD stage in Maya, however when publishing I would like to get the edits on the set layer to add to an override for a shot USD for example without needing to overwrite the set USD. We previously did this by querying the eidts with the getEditsThatAffect method in the MayaAPI, however that requires an assembly node as a parameter and does not work with a USD proxy shape.

If there was some way to list the edits to a USD stage in a similar way that this getEditsThatAffect method returns, that would be amazing to know.

Cheers

**Reply by Lucas3Dspain** [reply]
Hi @jonahjnewton ,
I'm not sure if this will help you but I guess you could set the anonymous leyer as the targetLayer (and do the work there)  then for publishing you save the anonymous layer. As the new version for x department

**Reply by gelshoffNV** [reply]
If you want to query / edit USD, you can gain access using the `mayaUsd.lib` like:
```
import mayaUsd.lib
stage = mayaUsd.lib.GetPrim('usdProxyShape').GetStage()
print(stage.GetSessionLayer().ExportToString())
```
where `'usdProxyShape'` is the dagPath to your stage proxy shape
